test/unit/lib/cms/aws_credentials_test.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Aws::AwsCredentialsTest < ActiveSupport::TestCase
+  setup do
+    CMS::AwsCredentials.instance.instance_variable_set('@credentials', nil)
+  end
+
+  teardown do
+    CMS::AwsCredentials.instance.instance_variable_set('@credentials', nil)
+  end
+
+  test 'returns IAM credentials when available' do
+    iam_params!
+    CMS::AwsCredentials.instance.expects(:sts_credentials).never
+
+    assert_equal iam_auth_params, CMS::AwsCredentials.instance.credentials
+  end
+
+  test 'returns STS credentials when available' do
+    sts_params!
+    CMS::AwsCredentials.instance.expects(:sts_credentials).returns(sts_credentials)
+
+    assert_equal({ credentials: sts_credentials }, CMS::AwsCredentials.instance.credentials)
+  end
+
+  test 'returns IAM credentials when both authentication types are available' do
+    full_params!
+    CMS::AwsCredentials.instance.expects(:sts_credentials).never
+
+    assert_equal iam_auth_params, CMS::AwsCredentials.instance.credentials
+  end
+
+  test 'raises an error if not enough credential params are present' do
+    CMS::S3.stubs(:config).returns(full_params.except(:role_arn, :access_key_id))
+
+    assert_raises(CMS::AwsCredentials::AuthenticationTypeError) do
+      CMS::AwsCredentials.instance.credentials
+    end
+  end
+
+  test 'is a singleton' do
+    assert CMS::AwsCredentials.ancestors.include?(Singleton), true
+  end
+
+  test 'raises an error if the web_identity_token_file does not exist' do
+    sts_params!
+
+    assert_raises(CMS::AwsCredentials::TokenNotFoundError) do
+      CMS::AwsCredentials.instance.credentials
+    end
+  end
+
+  test 'calls AWS to get an instance of STS credentials' do
```

Comment:
do we need this test at all? Why would we care whether AWS is called or not?

Reply:
Original test was essentially this
```ruby
    Aws::AssumeRoleWebIdentityCredentials.expects(:new).with(sts_auth_params).returns(assume_role_response)
```
It asserted that `AssumeRoleWebIdentityCredentials::new` was called properly (that's good, see below). I can't understand though why it had to return `Aws::STS::Types::AssumeRoleWithWebIdentityResponse` as this is a very different class. Neither I understand how exactly are we concerned about such a Response.

To implement the test I used the aws sdk client stubbing, as you can see and check that there is an actual call being performed. It is possible to check exact http call parameters but I consider this messing up with library internals that is more troubles than any benefit. Whether http call is immediately performed or lazily prformed later is also something library internal I don't care about.

That's why I ask whether we need this specific test.

P.S. The useful part for me is `Aws::AssumeRoleWebIdentityCredentials.expects(:new).with(sts_auth_params)` which is tested in the `STS credentials are created with correct parameters` test.

Reply:
I think I see the misunderstanding now.
The "calls AWS" in the sense "We perform a request specific to AWS" is not what we want to test at all, as it's part of the gem's internals. We only need to check that Porta's code is calling the correct method with the correct arguments, which's what `    Aws::AssumeRoleWebIdentityCredentials.expects(:new).with(sts_auth_params).returns(assume_role_response)` did/does.

> I can't understand though why it had to return

It doesn't have to anymore, since you enabled the "aws sdk client stubbing", which I had no idea existed in the first place.

So I think we're free to remove this test.